(0.104.0) Reformulate hydrostatic model timestepping

simone-silvestri opened the pull request:
This is a work-in-progress PR that aims to reformulate the timestepping scheme of the `HydrostaticFreeSurfaceModel` to 
make sure that three numerical properties are satisfied:

- tracers are globally conserved (the integral of a tracer field remains constant in the absence of forcings)
- tracers are locally conserved (an initially constant tracer remains constant in the absence of forcings)
- The free surface is evolved consistently with the grid thickness.

This is easier to achieve with an explicit discretization of the free surface (`ExplicitFreeSurface` and `SplitExplicitFreeSurface`) rather than an implicit free surface discretization and with RK3 rather than AB2.
However, this PR reformulates much of the timestepping algorithm, so it will be a lengthy WIP that will need to be thoroughly validated for all combinations of barotropic and baroclinic timestepping schemes.

More changes are performed to correctly abstract the timestepping schemes, which, at the moment, are tailored specifically to the `NonhydrostaticModel` (for example, the timesteppers include a `compute_pressure_correction!` or a ` correct_velocities` function, which are meaningful only for a non-hydrostatic Navier Stokes solver)

Edit: another objective I have is to remove any timestepping from the `update_state!` function. We should be able to call `update_state!` multiple times in a row without problems and timestepping inside `update_state!` negates this possibility.

TODO:
- [ ] <s> Remove timestepping from `update_state!` </s> In another PR
- [x] Clean up the time-stepping code
- [x] Introduce transport weights for transport averaging in the SplitExplicitFreeSurface
- [x] Implement a conservative, low-storage Runge Kutta method (of arbitrary order) using the Explicit free surface
- [x] Implement a conservative, low-storage Runge Kutta method (of arbitrary order) coupled with the SplitExplicitFreeSurface
- [x] Implement a conservative, low-storage Runge Kutta method (of arbitrary order) using the (PCG) ImplicitFreeSurface
- [x] Ensure that the conservation is satisfied on `DistributedGrids`
- [ ] <s> Implement a semi - conservative method for AB2  </s> In another PR
- [ ] <s> Address precision errors that tend to destroy conservation for constant fields </s> In another PR

simone-silvestri commented on src/Models/HydrostaticFreeSurfaceModels/HydrostaticFreeSurfaceModels.jl:
```diff
@@ -19,18 +19,50 @@ import Oceananigans: fields, prognostic_fields, initialize!
+fill_halo_regions!(ztype::ZCoordinate; kwargs...) = nothing
+
+struct ZStarCoordinate{CC} #, FC, CF, FF, H}
     storage :: CC # Storage space used in different ways by different timestepping schemes.
```
I think I will remove these fields, it was a way to have the grid metrics that are defined in the `Grids` module as fields so that we can use the boundary conditions infrastructure. However, BCs are not that trivial, so I don't think this is needed.

glwagner commented on src/Advection/reconstruction_coefficients.jl:
```diff
@@ -51,7 +51,7 @@ On a uniform `grid`, the coefficients are independent of the `xr` and `xi` value
         for j in 0:order-1
             for m in j+1:order
-                numerator   = sum(num_prod(i, m, l, r, xr, xi, shift, op, order, der) for l=0:order if l != m)
+                numerator   = sum(num_prod(i, m, l, r, FT.(xr), FT.(xi), shift, op, order, der) for l=0:order if l != m)
```
```suggestion
                xr = FT.(xr)
                xi = FT.(xi)
                numerator   = sum(num_prod(i, m, l, r, xr, xi, shift, op, order, der) for l=0:order if l != m)
```

glwagner commented on ext/OceananigansNCDatasetsExt.jl:
```diff
@@ -1352,6 +1354,12 @@ Write output to netcdf file `output_writer.filepath` at specified intervals. Inc
+    # Synchronize model state if needed
+    for field in fields(model)
+        synchronize_communication!(field)
+    end
```
how related to this PR?

glwagner commented on src/Advection/Advection.jl:
```diff
@@ -52,11 +54,11 @@ abstract type AbstractUpwindBiasedAdvectionScheme{B, FT} <: AbstractAdvectionSch
 # are constructed via metaprogramming. (The `advection_buffer` is the width of
 # the halo region required for an advection scheme on a non-immersed-boundary grid.)
 # An upper limit of `advection_buffer = 6` means we can build advection schemes up to
-# `Centered(order=12`) and `UpwindBiased(order=11)`. The list can be extended in order to
+# `Centered(order=10`) and `UpwindBiased(order=9)`. The list can be extended in order to
 # compile schemes with higher orders; for example `advection_buffers = [1, 2, 3, 4, 5, 6, 8]`
 # will compile schemes for `advection_buffer=8` and thus `Centered(order=16)` and `UpwindBiased(order=15)`.
 # Note that it is not possible to compile schemes for `advection_buffer = 41` or higher.
-const advection_buffers = [1, 2, 3, 4, 5, 6]
+const advection_buffers = [1, 2, 3, 4, 5]
```
why?

glwagner replied:
we should compile _past_ the recommended order. If order=9 is recommended we need order=11 to verify this.

simone-silvestri replied:
Okay, I guess I'll take this suggestion for https://github.com/CliMA/Oceananigans.jl/pull/4434, which, apart from this change, should be ready.

glwagner commented on src/Models/HydrostaticFreeSurfaceModels/pcg_implicit_free_surface_solver.jl:
```diff
@@ -16,17 +16,18 @@ The preconditioned conjugate gradient iterative implicit free-surface solver.
     architecture(solver.preconditioned_conjugate_gradient_solver)
 
+# The assumption is that the horizontal spacings do not depend on the z-direction
+@inline integrated_x_area(i, j, k, grid) = column_depthᶠᶜᵃ(i, j, grid) * Δyᶠᶜᵃ(i, j, grid.Nz, grid)
+@inline integrated_y_area(i, j, k, grid) = column_depthᶜᶠᵃ(i, j, grid) * Δxᶜᶠᵃ(i, j, grid.Nz, grid)
```
```suggestion
@inline integrated_x_area(i, j, k, grid) = column_depthᶠᶜᵃ(i, j, grid) * Δyᶠᶜᵃ(i, j, 1, grid)
@inline integrated_y_area(i, j, k, grid) = column_depthᶜᶠᵃ(i, j, grid) * Δxᶜᶠᵃ(i, j, 1, grid)
```

glwagner commented:
It's really hard for me to understand how this is an improvement. I think some explanation is needed (maybe docs or a docstring...)

simone-silvestri commented on src/TimeSteppers/runge_kutta_3.jl:
```diff
@@ -115,40 +115,32 @@ function time_step!(model::AbstractModel{<:RungeKutta3TimeStepper}, Δt; callbac
     # First stage
     #
 
-    compute_flux_bc_tendencies!(model)
```
I tried putting this back inside the `compute_tendencies!` given that these cannot be computed in `update_state!` anymore. I will verify everything is consistent.

glwagner commented:
> > why are there so many changes to WENO advection, etc?
> 
> I have merged in #4434 because I was expecting to merge it before this one. Maybe that was a mistake

Do you want us to also review the WENO changes then or ignore them?

simone-silvestri commented:
Yeah, also some of the code here probably still needs to change as I am fixing the implementation. 
My master plan is to get everything working properly with the numerics correct for all cases (all free surfaces, distributed, immersed boundaries etc...) so that I have a clear idea of all the changes to be done. Then, if I can keep it simple, go for one PR, while if the mole of the changes is too much for one PR, split the work in more PRs.
The advection PR is ready to review, though. I have added back the high-order advection schemes.

glwagner commented on src/Advection/weno_reconstruction.jl:
```diff
@@ -76,7 +76,7 @@ WENO{5, Float64, Float32}(order=9)
 function WENO(FT::DataType=Oceananigans.defaults.FloatType, FT2::DataType=Float32;
               order = 5,
-              buffer_scheme = nothing,
+              buffer_scheme = DescreasingOrderAdvectionScheme(),
```
```suggestion
              buffer_scheme = DecreasingOrderAdvectionScheme(),
```

simone-silvestri commented on src/OrthogonalSphericalShellGrids/distributed_zipper.jl:
```diff
@@ -6,7 +6,7 @@ using Oceananigans.DistributedComputations: cooperative_waitall!,
                                             loc_id
 
-using Oceananigans.Fields: location
+using Oceananigans.Fields: location, instantiated_location
```
thanks :)

giordano commented on src/TimeSteppers/runge_kutta_3.jl:
Stale imports
```suggestion
```
😇

glwagner commented on src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_rk_step.jl:
```diff
@@ -0,0 +1,166 @@
+    @apply_regionally begin
+        # Computing tendencies...
+        compute_momentum_tendencies!(model, callbacks)
+        compute_tracer_tendencies!(model)
```
These functions should receive the time-step

navidcy commented:
@simone-silvestri :

1. why is this PR breaking? It changes numerics or the API (or something else)?
2. why CI is broken? Is it expected?

navidcy commented:
I can try to push/review this!

glwagner commented on src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_ab2_step.jl:
```diff
@@ -117,8 +228,8 @@ end
-    α = convert(FT, 1.5) + χ
-    β = convert(FT, 0.5) + χ
+    α = 3*one(FT)/2 + χ
+    β = 1*one(FT)/2 + χ
```
or even just ``1//2 + χ`` actually

simone-silvestri replied:
This would work, I think the `//` has performance problems on GPUs

milankl replied:
```
1*one(FT)/2 + χ
1//2 + χ
```

are compiled to the same LLVM code

```julia
julia> f(x) = 1*one(eltype(x))/2 + x
f (generic function with 2 methods)

julia> @code_llvm f(1f0)
; Function Signature: f(Float32)
;  @ REPL[37]:1 within `f`
define float @julia_f_10935(float %"x::Float32") #0 {
top:
; ┌ @ float.jl:491 within `+`
   %0 = fadd float %"x::Float32", 5.000000e-01
   ret float %0
; └
}
```

vs

```julia
julia> f(x) = 1//2 + x
f (generic function with 2 methods)

julia> @code_llvm f(1f0)
; Function Signature: f(Float32)
;  @ REPL[32]:1 within `f`
define float @julia_f_10857(float %"x::Float32") #0 {
top:
; ┌ @ promotion.jl:429 within `+` @ float.jl:491
   %0 = fadd float %"x::Float32", 5.000000e-01
   ret float %0
; └
}
```

not sure why AMD would be different?

simone-silvestri replied:
I think that GPUCompiler rewrites a lot of the julia compilation to target different architectures so I do not expect `@code_llvm` to have the same output of a compiled kernel.

simone-silvestri commented:
On it! :)

giordano commented:
I had something similar in https://github.com/CliMA/Oceananigans.jl/pull/5105#issuecomment-3716441979, I suspect it may depend on whether some pkgextensions introduce ambiguities beyond those in Oceananigans itself, and whether they're loaded when you run the test. But I'd look like this PR is introducing ambiguities.

giordano commented:
On `main`:
```
julia> for mod in filter(name -> getproperty(Oceananigans, name) isa Module, names(Oceananigans; all=true))
           n = length(Test.detect_ambiguities(getproperty(Oceananigans, mod); recursive=true))
           n > 0 && @info "$(mod) has $(n) ambiguities"
       end
[ Info: AbstractOperations has 100 ambiguities
[ Info: Advection has 2 ambiguities
[ Info: BoundaryConditions has 84 ambiguities
[ Info: Diagnostics has 1 ambiguities
[ Info: DistributedComputations has 24 ambiguities
[ Info: Fields has 46 ambiguities
[ Info: Grids has 14 ambiguities
[ Info: ImmersedBoundaries has 23 ambiguities
[ Info: Models has 10 ambiguities
[ Info: MultiRegion has 31 ambiguities
[ Info: Oceananigans has 349 ambiguities
[ Info: Operators has 21 ambiguities
[ Info: OrthogonalSphericalShellGrids has 4 ambiguities
[ Info: OutputReaders has 10 ambiguities
[ Info: OutputWriters has 6 ambiguities
[ Info: Solvers has 5 ambiguities
[ Info: TurbulenceClosures has 33 ambiguities
[ Info: Utils has 2 ambiguities
```
on this PR:
```
julia> for mod in filter(name -> getproperty(Oceananigans, name) isa Module, names(Oceananigans; all=true))
           n = length(Test.detect_ambiguities(getproperty(Oceananigans, mod); recursive=true))
           n > 0 && @info "$(mod) has $(n) ambiguities"
       end
[ Info: AbstractOperations has 104 ambiguities
[ Info: Advection has 2 ambiguities
[ Info: BoundaryConditions has 84 ambiguities
[ Info: Diagnostics has 1 ambiguities
[ Info: DistributedComputations has 24 ambiguities
[ Info: Fields has 46 ambiguities
[ Info: Grids has 14 ambiguities
[ Info: ImmersedBoundaries has 23 ambiguities
[ Info: Models has 11 ambiguities
[ Info: MultiRegion has 31 ambiguities
[ Info: Oceananigans has 354 ambiguities
[ Info: Operators has 21 ambiguities
[ Info: OrthogonalSphericalShellGrids has 4 ambiguities
[ Info: OutputReaders has 10 ambiguities
[ Info: OutputWriters has 6 ambiguities
[ Info: Solvers has 5 ambiguities
[ Info: TurbulenceClosures has 33 ambiguities
[ Info: Utils has 2 ambiguities
```
so there are 4 new ambiguities in `AbstractOperations` and 1 new one in `Models`.  The one in `Models` is somewhat easy to find, on `main`:
```
julia> Test.detect_ambiguities(Oceananigans.Models; recursive=true)
10-element Vector{Tuple{Method, Method}}:
 (flattened_node(::Any, grid::Oceananigans.Grids.AbstractGrid{<:Any, <:Any, Flat, Flat}) @ Oceananigans.Models.LagrangianParticleTracking ~/.julia/packages/Oceananigans/eM3fS/src/Models/LagrangianParticleTracking/LagrangianParticleTracking.jl:127, flattened_node(::Any, grid::Oceananigans.Grids.AbstractGrid{<:Any, Flat, Flat}) @ Oceananigans.Models.LagrangianParticleTracking ~/.julia/packages/Oceananigans/eM3fS/src/Models/LagrangianParticleTracking/LagrangianParticleTracking.jl:129)
 (sum_of_velocities(U1::PrescribedVelocityFields, U2, U3) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/eM3fS/src/Models/HydrostaticFreeSurfaceModels/prescribed_hydrostatic_velocity_fields.jl:86, sum_of_velocities(U1, U2, U3::PrescribedVelocityFields) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/eM3fS/src/Models/HydrostaticFreeSurfaceModels/prescribed_hydrostatic_velocity_fields.jl:88)
 (split_explicit_substepping(::Nothing, substeps, fixed_Δt, grid, averaging_kernel, gravitational_acceleration) @ Oceananigans.Models.HydrostaticFreeSurfaceModels.SplitExplicitFreeSurfaces ~/.julia/packages/Oceananigans/eM3fS/src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_free_surface.jl:156, split_explicit_substepping(cfl, ::Nothing, fixed_Δt, grid, averaging_kernel, gravitational_acceleration) @ Oceananigans.Models.HydrostaticFreeSurfaceModels.SplitExplicitFreeSurfaces ~/.julia/packages/Oceananigans/eM3fS/src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_free_surface.jl:169)
 (sum_of_velocities(U1, U2::PrescribedVelocityFields, U3) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/eM3fS/src/Models/HydrostaticFreeSurfaceModels/prescribed_hydrostatic_velocity_fields.jl:87, sum_of_velocities(U1, U2, U3::PrescribedVelocityFields) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/eM3fS/src/Models/HydrostaticFreeSurfaceModels/prescribed_hydrostatic_velocity_fields.jl:88)
 (hydrostatic_tendency_fields(velocities, free_surface::ExplicitFreeSurface, grid, tracer_names, bcs) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/eM3fS/src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_field_tuples.jl:15, hydrostatic_tendency_fields(::PrescribedVelocityFields, free_surface, grid, tracer_names, bcs) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/eM3fS/src/Models/HydrostaticFreeSurfaceModels/prescribed_hydrostatic_velocity_fields.jl:70)
 (hydrostatic_tendency_fields(velocities, free_surface::SplitExplicitFreeSurface, grid, tracer_names, bcs) @ Oceananigans.Models.HydrostaticFreeSurfaceModels.SplitExplicitFreeSurfaces ~/.julia/packages/Oceananigans/eM3fS/src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_free_surface.jl:197, hydrostatic_tendency_fields(::PrescribedVelocityFields, free_surface, grid, tracer_names, bcs) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/eM3fS/src/Models/HydrostaticFreeSurfaceModels/prescribed_hydrostatic_velocity_fields.jl:70)
 (sum_of_velocities(U1::PrescribedVelocityFields, U2, U3) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/eM3fS/src/Models/HydrostaticFreeSurfaceModels/prescribed_hydrostatic_velocity_fields.jl:86, sum_of_velocities(U1, U2::PrescribedVelocityFields, U3) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/eM3fS/src/Models/HydrostaticFreeSurfaceModels/prescribed_hydrostatic_velocity_fields.jl:87)
 (flattened_node(::Any, grid::Oceananigans.Grids.AbstractGrid{<:Any, Flat, <:Any, Flat}) @ Oceananigans.Models.LagrangianParticleTracking ~/.julia/packages/Oceananigans/eM3fS/src/Models/LagrangianParticleTracking/LagrangianParticleTracking.jl:128, flattened_node(::Any, grid::Oceananigans.Grids.AbstractGrid{<:Any, Flat, Flat}) @ Oceananigans.Models.LagrangianParticleTracking ~/.julia/packages/Oceananigans/eM3fS/src/Models/LagrangianParticleTracking/LagrangianParticleTracking.jl:129)
 (sum_of_velocities(U1::PrescribedVelocityFields, U2) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/eM3fS/src/Models/HydrostaticFreeSurfaceModels/prescribed_hydrostatic_velocity_fields.jl:83, sum_of_velocities(U1, U2::PrescribedVelocityFields) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/eM3fS/src/Models/HydrostaticFreeSurfaceModels/prescribed_hydrostatic_velocity_fields.jl:84)
 (flattened_node(::Any, grid::Oceananigans.Grids.AbstractGrid{<:Any, <:Any, Flat, Flat}) @ Oceananigans.Models.LagrangianParticleTracking ~/.julia/packages/Oceananigans/eM3fS/src/Models/LagrangianParticleTracking/LagrangianParticleTracking.jl:127, flattened_node(::Any, grid::Oceananigans.Grids.AbstractGrid{<:Any, Flat, <:Any, Flat}) @ Oceananigans.Models.LagrangianParticleTracking ~/.julia/packages/Oceananigans/eM3fS/src/Models/LagrangianParticleTracking/LagrangianParticleTracking.jl:128)
```
in this PR:
```
julia> Test.detect_ambiguities(Oceananigans.Models; recursive=true)
11-element Vector{Tuple{Method, Method}}:
 (flattened_node(::Any, grid::Oceananigans.Grids.AbstractGrid{<:Any, <:Any, Flat, Flat}) @ Oceananigans.Models.LagrangianParticleTracking ~/.julia/packages/Oceananigans/S0uCG/src/Models/LagrangianParticleTracking/LagrangianParticleTracking.jl:127, flattened_node(::Any, grid::Oceananigans.Grids.AbstractGrid{<:Any, Flat, <:Any, Flat}) @ Oceananigans.Models.LagrangianParticleTracking ~/.julia/packages/Oceananigans/S0uCG/src/Models/LagrangianParticleTracking/LagrangianParticleTracking.jl:128)
 (flattened_node(::Any, grid::Oceananigans.Grids.AbstractGrid{<:Any, <:Any, Flat, Flat}) @ Oceananigans.Models.LagrangianParticleTracking ~/.julia/packages/Oceananigans/S0uCG/src/Models/LagrangianParticleTracking/LagrangianParticleTracking.jl:127, flattened_node(::Any, grid::Oceananigans.Grids.AbstractGrid{<:Any, Flat, Flat}) @ Oceananigans.Models.LagrangianParticleTracking ~/.julia/packages/Oceananigans/S0uCG/src/Models/LagrangianParticleTracking/LagrangianParticleTracking.jl:129)
 (split_explicit_substepping(::Nothing, substeps, fixed_Δt, grid, averaging_kernel, gravitational_acceleration) @ Oceananigans.Models.HydrostaticFreeSurfaceModels.SplitExplicitFreeSurfaces ~/.julia/packages/Oceananigans/S0uCG/src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_free_surface.jl:156, split_explicit_substepping(cfl, ::Nothing, fixed_Δt, grid, averaging_kernel, gravitational_acceleration) @ Oceananigans.Models.HydrostaticFreeSurfaceModels.SplitExplicitFreeSurfaces ~/.julia/packages/Oceananigans/S0uCG/src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_free_surface.jl:169)
 (flattened_node(::Any, grid::Oceananigans.Grids.AbstractGrid{<:Any, Flat, <:Any, Flat}) @ Oceananigans.Models.LagrangianParticleTracking ~/.julia/packages/Oceananigans/S0uCG/src/Models/LagrangianParticleTracking/LagrangianParticleTracking.jl:128, flattened_node(::Any, grid::Oceananigans.Grids.AbstractGrid{<:Any, Flat, Flat}) @ Oceananigans.Models.LagrangianParticleTracking ~/.julia/packages/Oceananigans/S0uCG/src/Models/LagrangianParticleTracking/LagrangianParticleTracking.jl:129)
 (sum_of_velocities(U1::PrescribedVelocityFields, U2) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/S0uCG/src/Models/HydrostaticFreeSurfaceModels/prescribed_hydrostatic_velocity_fields.jl:83, sum_of_velocities(U1, U2::PrescribedVelocityFields) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/S0uCG/src/Models/HydrostaticFreeSurfaceModels/prescribed_hydrostatic_velocity_fields.jl:84)
 (hydrostatic_tendency_fields(velocities, free_surface::SplitExplicitFreeSurface, grid, tracer_names, bcs) @ Oceananigans.Models.HydrostaticFreeSurfaceModels.SplitExplicitFreeSurfaces ~/.julia/packages/Oceananigans/S0uCG/src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_free_surface.jl:197, hydrostatic_tendency_fields(::PrescribedVelocityFields, free_surface, grid, tracer_names, bcs) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/S0uCG/src/Models/HydrostaticFreeSurfaceModels/prescribed_hydrostatic_velocity_fields.jl:70)
 (ab2_step!(model, ::ImplicitFreeSurface, grid, Δt, callbacks) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/S0uCG/src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_ab2_step.jl:84, ab2_step!(model::NonhydrostaticModel, args...) @ Oceananigans.Models.NonhydrostaticModels ~/.julia/packages/Oceananigans/S0uCG/src/Models/NonhydrostaticModels/nonhydrostatic_ab2_step.jl:10)
 (sum_of_velocities(U1, U2::PrescribedVelocityFields, U3) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/S0uCG/src/Models/HydrostaticFreeSurfaceModels/prescribed_hydrostatic_velocity_fields.jl:87, sum_of_velocities(U1, U2, U3::PrescribedVelocityFields) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/S0uCG/src/Models/HydrostaticFreeSurfaceModels/prescribed_hydrostatic_velocity_fields.jl:88)
 (sum_of_velocities(U1::PrescribedVelocityFields, U2, U3) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/S0uCG/src/Models/HydrostaticFreeSurfaceModels/prescribed_hydrostatic_velocity_fields.jl:86, sum_of_velocities(U1, U2::PrescribedVelocityFields, U3) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/S0uCG/src/Models/HydrostaticFreeSurfaceModels/prescribed_hydrostatic_velocity_fields.jl:87)
 (sum_of_velocities(U1::PrescribedVelocityFields, U2, U3) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/S0uCG/src/Models/HydrostaticFreeSurfaceModels/prescribed_hydrostatic_velocity_fields.jl:86, sum_of_velocities(U1, U2, U3::PrescribedVelocityFields) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/S0uCG/src/Models/HydrostaticFreeSurfaceModels/prescribed_hydrostatic_velocity_fields.jl:88)
 (hydrostatic_tendency_fields(velocities, free_surface::ExplicitFreeSurface, grid, tracer_names, bcs) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/S0uCG/src/Models/HydrostaticFreeSurfaceModels/explicit_free_surface.jl:43, hydrostatic_tendency_fields(::PrescribedVelocityFields, free_surface, grid, tracer_names, bcs) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/S0uCG/src/Models/HydrostaticFreeSurfaceModels/prescribed_hydrostatic_velocity_fields.jl:70)
```
The one which is only in this PR is
```
 (ab2_step!(model, ::ImplicitFreeSurface, grid, Δt, callbacks) @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/.julia/packages/Oceananigans/S0uCG/src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_ab2_step.jl:84, ab2_step!(model::NonhydrostaticModel, args...) @ Oceananigans.Models.NonhydrostaticModels ~/.julia/packages/Oceananigans/S0uCG/src/Models/NonhydrostaticModels/nonhydrostatic_ab2_step.jl:10)
```

simone-silvestri commented on src/Models/NonhydrostaticModels/nonhydrostatic_ab2_step.jl:
```diff
@@ -0,0 +1,68 @@
+Dispatches to `pressure_correction_ab2_step!` which implements a predictor-corrector scheme
+"""
+ab2_step!(model::NonhydrostaticModel, args...) = 
+    pressure_correction_ab2_step!(model, args...)
```
I am not sure if I necessarily fixed the correct ambiguities so just in case I tryed to fix more than the extra 5 (10 in total) so that at least this PR is a positive change

simone-silvestri commented on src/Fields/field_tuples.jl:
```diff
@@ -56,17 +56,13 @@ Fill halo regions for all `fields`. The algorithm:
-    for field in fields
-        fill_halo_regions!(field, args...; kwargs...)
+    for i in eachindex(fields)
+        @inbounds fill_halo_regions!(fields[i], args...; kwargs...)
```
cthulhu was giving me type instability for `for field in fields`

simone-silvestri commented:
Everything passes, this PR should be ready to review and merge. It also changes much of the underlying infrastructure so it will need to be followed by some PRs that fix the checkpointing.

simone-silvestri commented:
> > It also changes much of the underlying infrastructure so it will need to be followed by some PRs that fix the checkpointing.
> 
> Checkpointing is already being revamped in #4892, so there'll probably a fair amount of conflicts between the two PRs. Should we merge #4892 first? It seems simplify the checkpointing infra-structure so I'm assuming conflicts will be easier to fix on this side. @ali-ramadhan thoughts?

Yeash there are a fair bit of conflicts. The checkpointing for zstar needs to change, all the work done for CATKE for example can be reverted after this PR. But I guess let's merge that one first then I can correct everything. The important thing is not to let this one hang too much since it is already playing catch up and has important changes that can simplify the internals quite a bit.

ali-ramadhan commented:
> 707 commits

:open_mouth:

simone-silvestri commented:
> > 707 commits
> 
> 😮

Yeah it has been brewing for a while 😅

ali-ramadhan commented on src/TimeSteppers/split_runge_kutta.jl:
```diff
@@ -0,0 +1,201 @@
+    end
+    if coefficients == nothing 
+        coefficients = tuple(collect(stages:-1:1)...)
+    end
```
By default `coefficients = nothing` and `stages = nothing` so won't this error then?

```julia
julia> nothing:-1:1
ERROR: MethodError: no method matching (::Colon)(::Nothing, ::Int64, ::Int64)
```

ali-ramadhan commented on src/TimeSteppers/split_runge_kutta.jl:
```diff
@@ -0,0 +1,201 @@
+References
+==========
+
+To be added once the paper is submitted...
```
🚀

ali-ramadhan commented on src/Models/HydrostaticFreeSurfaceModels/barotropic_pressure_correction.jl:
I know this isn't part of the PR but are these derivatives operators correct? η is at ccf and we want to go to fcc (where u lives). So we need both a derivative and an interpolation?

simone-silvestri replied:
hmmm, we probably need a two-dimensional operator here since `η` is two-dimensional. This is a problem also in #5101

ali-ramadhan commented on src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/barotropic_split_explicit_corrector.jl:
```diff
@@ -60,3 +80,59 @@ end
+
+    @inbounds begin
+        ũ⁺ = u[i, j, k] + (Ũ[i, j, 1] - U̅[i, j, 1]) / Hᶠᶜ
+        ṽ⁺ = v[i, j, k] + (Ṽ[i, j, 1] - V̅[i, j, 1]) / Hᶜᶠ
```
This might introduce NaNs in the halo regions? See: https://github.com/CliMA/Oceananigans.jl/issues/5024

Might be good to check if `Hᶠᶜ` or `Hᶜᶠ` are zero, and set `ũ⁺ = 0` or `ṽ⁺ = 0` appropriately?

ali-ramadhan commented on src/TimeSteppers/split_runge_kutta.jl:
```diff
@@ -0,0 +1,201 @@
+        error("Cannot specify both `coefficients` and `stages`.")
+    end
+    if coefficients == nothing 
+        coefficients = tuple(collect(stages:-1:1)...)
```
```suggestion
    if !isnothing(coefficients) && !isnothing(stages)
        error("Cannot specify both `coefficients` and `stages`.")
    end
    if isnothing(coefficients)
        coefficients = tuple(collect(stages:-1:1)...)
```

ali-ramadhan replied:
I'm not sure if it matters but checking for nothingness in Julia seems to be done with the `===` operator but there's also an `isnothing` function that might be good to use so Julia can check for nothingness in the best way (especially if it changes in the future).

```
help?> isnothing
search: isnothing nothing Nothing Base.notnothing @something string something Cstring ismissing

  isnothing(x)

  Return true if x === nothing, and return false if not.

  │ Julia 1.1
  │
  │  This function requires at least Julia 1.1.

  See also something, Base.notnothing, ismissing.
```

ali-ramadhan commented on src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_rk_step.jl:
```diff
@@ -0,0 +1,222 @@
+            Gⁿ = model.timestepper.Gⁿ[tracer_name]
+            Ψ⁻ = model.timestepper.Ψ⁻[tracer_name]
+            c  = tracers[tracer_name]
+            closure = model.closure
```
```suggestion
```

No need for this since we have closure = model.closure outside the loop.

ali-ramadhan commented on src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_ab2_step.jl:
Also no need for this since we have `closure = model.closure` outside the loop.

ali-ramadhan reviewed (commented):
This is awesome! Excited to see if our simulations at atdepth will look better with this in v0.104.

I do agree with @glwagner that it would be good to have some documentation, both to advise users on which timestepper to use (or at least describe them at a high-level so users can make an informed decision) and to describe what the timesteppers do in detail. 

I left some random comments here and there (and a few more general ones below). I did not manage to look through everything (99 files changed!) but will try to revisit.

A few comments:

1. I'm seeing several `TODO` comments scattered throughout the codebase (some existing, some new). I don't know if we ever discussed this (for Oceananigans community development) but would a good approach to documenting TODO's be to open GitHub issues for each TODO? Or at least for major ones worth documenting.
2. I'm seeing tests for `SplitRungeKuttaN` up to N = 5 but up to N = 40 is supported. Might be good to test at least up to 10 or 20 if it's not too expensive for the compiler. If the Julia compiler chokes at higher N then that might provide a good upper bound for N.
3. Do we expect `ShallowWaterModel` to work with `:SplitRungeKuttaN`? I think so but I don't think we have a test for this?

simone-silvestri commented:
> This is awesome! Excited to see if our simulations at atdepth will look better with this in v0.104.

Thanks for the review! I will address all the comments by tomorrow.

> I do agree with @glwagner that it would be good to have some documentation, both to advise users on which timestepper to use (or at least describe them at a high-level so users can make an informed decision) and to describe what the timesteppers do in detail.

I can whipe out a doc page for the timesteppers, do you think it should be part of this PR?

> I left some random comments here and there (and a few more general ones below). I did not manage to look through everything (99 files changed!) but will try to revisit.
> 
> A few comments:
> 
> 1. I'm seeing several `TODO` comments scattered throughout the codebase (some existing, some new). I don't know if we ever discussed this but would a good approach to documenting TODO's be to open GitHub issues for each TODO? Or at least for major ones worth documenting.
> 2. I'm seeing tests for `SplitRungeKuttaN` up to N = 5 but up to N = 40 is supported. Might be good to test at least up to 10 or 20 if it's not too expensive for the compiler. If the Julia compiler chokes at higher N then that might provide a good upper bound for N.
> 3. Do we expect `ShallowWaterModel` to work with `:SplitRungeKuttaN`? I think so but I don't think we have a test for this?

1. Indeed issues are definitely a good idea, I was planning to open issues as soon as this is merged so we know where to go next.
2. Agreed, maybe we can cap it at 5, I don't think there is a good reason to go up to 40, if these high number of stages is required then maybe just decreasing the time-step might be a better idea 😅 
3. For the moment this timestepper works only for the hydrostaticfreesurfacemodel, the other two models use the `RungeKutta3` timestepper which has a different formulation, however, in principle it is not difficult to adapt other models to this formulation, if there is interest.

simone-silvestri commented on test/test_zstar_conservation.jl:
unfortunately this PR makes adams-bashforth non-conserving. it was conserving before for an incorrect reason, more work is required to make it actually conserving so I was thinking to do it after this PR (another issue to be opened :) ). ZCoordinate is never conserving because the velocity at the surface does not go to zero.

ali-ramadhan replied:
Makes sense, thanks for explaining!

ali-ramadhan commented on src/TimeSteppers/TimeSteppers.jl:
```diff
@@ -61,8 +58,15 @@ TimeStepper(::Val{:QuasiAdamsBashforth2}, args...; kwargs...) =
+for stages in 1:40
+    @eval TimeStepper(::Val{Symbol(:SplitRungeKutta, $stages)}, args...; kwargs...) =
+              SplitRungeKuttaTimeStepper(args...; coefficients=tuple(collect($stages:-1:1)...), kwargs...)
+end
```
Sounds good!

ali-ramadhan commented:
@simone-silvestri I agree that checkpointing in the middle of a stage might be weird so that makes sense to me! So right now if you want checkpointing  you cannot use split RK/RK3 and must use QAB2?

And updated docs are looking good!

navidcy commented on test/test_checkpointer.jl:
```diff
@@ -1815,14 +1817,16 @@ for arch in archs
+            @testset "CATKE closure checkpointing [$(typeof(arch)), $timestepper]" begin
+                @info "  Testing CATKE closure checkpointing [$(typeof(arch)), $timestepper]..."
+                test_checkpointing_catke_closure(arch, timestepper)
+            end
```
CATKE tests with QuasiAdamsBashforth2 pass if we change the atol from 0 to 1e-20.
Is there a reason we expect atol=0?

ali-ramadhan replied:
Nice catch! I started with `atol=0` on every test to see if I can get bit-for-bit reproducibility and it worked for most tests.

When I started testing RK3 `atol=0` was too tight, and I think we needed `atol=1e-20` due to floating point order of operations.

I'd like to keep `atol=0` where possible but `atol=1e-20` on CATKE QAB2 checkpointing tests sounds good and means it's working perfectly!

simone-silvestri replied:
yeah I think now we compute tendencies in `update_state!` also when building the model and that advances TKE unfortunately, which is something I would like to change.

simone-silvestri replied:
I think in CATKE checkpointing, `e` is not really tested also on main because the extrema are 
```
julia> extrema(new_simulation.model.tracers.e)
(1.7399248638853215e-26, 3.8061030807487524e-24)
```
and the atol is -20. It seems that the tests pass with `atol=0` for all variables except for `e`. Should I skip the tests only for `e`?

simone-silvestri replied:
Actually, if it is ok, I think the best strategy is to merge this and then cover up all the todos in later PRs, otherwise this PR will continue for ever given it changes much of the internals.

ali-ramadhan commented:
Happy to help fix the checkpointing tests! I can take a look tomorrow. But also happy to merge and I can fix them in a subsequent PR to avoid delaying this PR from getting merged.

simone-silvestri commented:
> For the shallow-water model, I think we are missing the `ab2_step!(model::ShallowWaterModel, Δt, callbacks)` and `cache_current_fields!(model::ShallowWaterModel)` methods

Thanks I will add these methods

giordano commented on src/TimeSteppers/store_tendencies.jl:
This file is empty.  Was it supposed to be deleted entirely?

navidcy commented on Project.toml:
```diff
@@ -10,6 +10,7 @@ CubedSphere = "7445602f-e544-4518-8976-18f8e8ae6cdb"
 Dates = "ade2ca70-3891-5945-98fb-dc099432e06a"
 Distances = "b4f34e82-e78d-54a5-968a-f98e89d6e8f7"
 DocStringExtensions = "ffbed154-4ef7-542d-bbb7-c09d3a79fcae"
+DoubleFloats = "497a8b3b-efae-58df-a0af-a86822472b78"
```
Is this related to this PR?

glwagner commented on docs/src/numerical_implementation/time_stepping.md:
```diff
@@ -1,4 +1,89 @@
+!!! note "Time stepper availability in Oceananigans models"
+    - `NonhydrostaticModel`: `QuasiAdamsBashforth2`, `RungeKutta3TimeStepper` (default)
+    - `ShallowWaterModel`: `QuasiAdamsBashforth2TimeStepper`, `RungeKutta3` (default)
+    - `HydrostaticFreeSurfaceModel`: `QuasiAdamsBashforth2TimeStepper` (default), `SplitRungeKuttaTimeStepper`
```
these docs are great, but there are no code examples. I think the code examples are actually more important than the theory (theory is cool, but irrelevant if you can't actually use the code)

glwagner commented on src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/barotropic_split_explicit_corrector.jl:
```diff
@@ -22,7 +29,20 @@ function compute_barotropic_mode!(U̅, V̅, grid, u, v)
+The correction ensures that the depth-integrated baroclinic velocity matches the
+filtered barotropic velocity from the split-explicit scheme:
+
+    u_corrected = u + (U_filtered - U_baroclinic) / H
```
elsewhere we have discussed how it isn't great to mix math and english notation... if we can define symbols for these variables, it might be preferred. presumably the paper won't use "_corrected" and "_filtered", right?

glwagner commented on src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_free_surface.jl:
```diff
@@ -178,6 +178,10 @@ end
 
+# Disambiguation
+split_explicit_substepping(::Nothing, ::Nothing, fixed_Δt, grid, averaging_kernel, gravitational_acceleration) =
+    split_explicit_substepping(nothing, MINIMUM_SUBSTEPS, fixed_Δt, grid, averaging_kernel, gravitational_acceleration)
```
what determines `MINIMUM_SUBSTEPS`?

glwagner commented on src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_timesteppers.jl:
```diff
@@ -10,22 +10,27 @@ A timestepping scheme used for substepping in the split-explicit free surface so
+##### AdamsBasforth3 timestepper
+#####
+
 struct AdamsBashforth3Scheme{CC, FC, CF, FT}
```
also --- it seems like the timesteppers for the hydrostatic model combine the splitting, eg the whole time-stepper is called "splitrungekutta". How is it possible to have a timestepper _just_ for the barotropic mode as here? does this actually work / is this correct?

simone-silvestri replied:
I can remove it, it is not really necessary and not used.

briochemc replied:
Oh oops sorry for the noise then!

glwagner commented on src/Models/HydrostaticFreeSurfaceModels/update_hydrostatic_free_surface_model_state.jl:
```diff
@@ -3,97 +3,121 @@ using Oceananigans.Biogeochemistry: update_biogeochemical_state!
 
-    compute_tendencies &&
-        @apply_regionally compute_tendencies!(model, callbacks)
+    @apply_regionally compute_momentum_tendencies!(model, callbacks)
```
What is the impact of this change on distributed scaling?

glwagner commented on src/Models/HydrostaticFreeSurfaceModels/z_star_vertical_spacing.jl:
```diff
@@ -13,155 +13,112 @@ import Oceananigans: prognostic_state, restore_prognostic_state!
+"""
+function rk_substep_grid!(grid::MutableGridOfSomeKind, model, ztype::ZStarCoordinate, Δt)
+    launch!(architecture(grid), grid, surface_kernel_parameters(grid), _update_zstar_scaling!, model.free_surface.displacement, grid)
+    if model.clock.stage == length(model.timestepper.β)
```
what is the significance of `length(β)`? this is the number of stages? please make the number of stages explicit.

glwagner commented on src/Models/NonhydrostaticModels/update_nonhydrostatic_model_state.jl:
```diff
@@ -32,31 +30,30 @@ function update_state!(model::NonhydrostaticModel, callbacks=[]; compute_tendenc
 
     # Calculate diffusivities and hydrostatic pressure
-    @apply_regionally compute_auxiliaries!(model)
+    compute_auxiliaries!(model)
```
so this PR removes support for multi region nonhydrostatic?

glwagner commented on src/Models/HydrostaticFreeSurfaceModels/update_hydrostatic_free_surface_model_state.jl:
```diff
@@ -3,97 +3,121 @@ using Oceananigans.Biogeochemistry: update_biogeochemical_state!
+    @apply_regionally begin
+        surface_params = surface_kernel_parameters(grid)
+        volume_params = volume_kernel_parameters(grid)
+        κ_params = diffusivity_kernel_parameters(grid)
```
why are these different?

simone-silvestri commented on src/Models/interleave_communication_and_computation.jl:
```diff
@@ -65,3 +65,41 @@ function interior_tendency_kernel_parameters(arch::AsynchronousDistributed, grid
+Return kernel parameters for computing 3D (volume) variables including halo regions.
+Similar to `surface_kernel_parameters` but for three-dimensional fields.
+"""
+@inline function volume_kernel_parameters(grid)
```
closures have their own halo requirement that dictate how many points in the interior are valid, so in practice we can compute diffusivity up to a certain number of cells and then we need to recompute diffusivity in those places. This is also why the buffer kernels need to be different.

simone-silvestri commented on src/TimeSteppers/TimeSteppers.jl:
```diff
@@ -61,8 +57,15 @@ TimeStepper(::Val{:QuasiAdamsBashforth2}, args...; kwargs...) =
+for stages in 2:5
+    @eval TimeStepper(::Val{Symbol(:SplitRungeKutta, $stages)}, args...; kwargs...) =
+              SplitRungeKuttaTimeStepper(args...; coefficients=tuple(collect($stages:-1:1)...), kwargs...)
+end
```
to be able to do 
```julia
timestepper = :SplitRungeKutta2 
```
of 
```julia
timestepper = :SplitRungeKutta4
```
we discussed above with @ali-ramadhan that we do not want to support schemes with higher number of stages since they are not tested and not be beneficial

simone-silvestri replied:
I can add something in the docs to say that we can build a timestepper appending the number of stages (:SplitRungeKutta3 for example).

giordano commented on src/Models/VarianceDissipationComputations/compute_dissipation.jl:
```diff
@@ -61,8 +61,8 @@ end
-function assemble_advective_dissipation!(P, grid, ::RungeKuttaScheme, substep, Fⁿ, Fⁿ⁻¹, Uⁿ, Uⁿ⁻¹, cⁿ⁺¹, cⁿ)
-    if substep == 3
+function assemble_advective_dissipation!(P, grid, ts::RungeKuttaScheme, substep, Fⁿ, Fⁿ⁻¹, Uⁿ, Uⁿ⁻¹, cⁿ⁺¹, cⁿ)
+    if substep == ts.Nstages
```
From JETLS:

> FieldError: type Oceananigans.TimeSteppers.RungeKutta3TimeStepper has no field `Nstages`, available fields: `γ¹`, `γ²`, `γ³`, `ζ²`, `ζ³`, `Gⁿ`, `G⁻`, `implicit_solver`

https://github.com/CliMA/Oceananigans.jl/blob/87c7213e59d924dcc02236edc25caab5b6f156c2/src/TimeSteppers/runge_kutta_3.jl#L16-L25

`RungeKuttaScheme` is https://github.com/CliMA/Oceananigans.jl/blob/87c7213e59d924dcc02236edc25caab5b6f156c2/src/Models/VarianceDissipationComputations/VarianceDissipationComputations.jl#L29 so this method effectively only works with `SplitRungeKuttaTimeStepper`: https://github.com/CliMA/Oceananigans.jl/blob/87c7213e59d924dcc02236edc25caab5b6f156c2/src/TimeSteppers/split_runge_kutta.jl#L14-L20

Same for `assemble_diffusive_dissipation!` below.
